### DISCUSSION TITLE
monorepo项目中，dev模式如何得知被关联的子项目首次dev编译完成？

### DISCUSSION THREAD

**Discussion by PopperLi** [umijs/father#731] [Q&A] [answered]
我有A、B、C三个包，A依赖B、C，在开发 A包时需要先等待B、C初次完成构建，但是dev是并行的，如何得知B、C两个包完成了首次构建？

**Comment by PeachScript** [comment]
暂时只能人肉判定，得实现 `onDevCompileDone` 的插件 API 才能支持

**Reply by PopperLi** [reply]
麻烦给个思路

**Reply by PeachScript** [reply, marked as answer]
需要三处改动：
1. 在 bundle 以及 bundless watch 模式构建完成时，通过 applyPlugins 执行插件钩子：https://github.com/umijs/umi/blob/13dd25b2b8e65a26764f796bb8be5aafb039be60/packages/preset-umi/src/commands/dev/dev.ts#L382-L385
2. 注册该插件钩子：https://github.com/umijs/father/blob/21006384af8be2bbb508a6548df2af979c266209/src/registerMethods.ts#L8
3. 提供对应类型：https://github.com/umijs/umi/blob/c5c292ee842673e2d46893ba5e364f13377db98a/packages/preset-umi/src/types.ts#L204-L209

参数暂时应该只需要 `isFirstCompile`

**Comment by fz6m** [comment]
如果是为了在 monorepo 中分包开发，可以使用 umi 4 的 `monorepoRedirect` 功能，支持直接修改 TS 源码直接热更新，无需编译多此一举。
